Try: Larger drag handles

chrisvanpatten opened the pull request:
This is an attempt to make a larger target area for drag handles. The handle itself doesn't change its appearance, but you now have a larger target area for your cursor, as depicted here:

![fullscreen_4_10_18__10_49](https://user-images.githubusercontent.com/1231306/46482346-6c0ad700-c7c3-11e8-94a2-7431b0d52d90.png)

Basically the visible handle is now a pseudo element positioned inside the larger handle container.

Fixes #8206 — the larger target area makes it a bit easier to grab the handle when there's a block below the spacer block.

I'd like to also extend this to the image block but just want the OK before going forward with that.

### Some other notes:

I introduced a new variable, `$resize-handler-size: 16px`, because until now the spacer and image drag handles were different sizes (15px and 16px respectively). If this approach is worth moving forward with, I'll use this variable on the image drag handles as well.

I am also using another new variable `$resize-handler-container-size` to calculate the size of the container (the drag handle plus padding). I'm doing it inline in the spacer block drag handle CSS block but if this moves forward I would move it in with the other variables. This is just temporary :)

chrisvanpatten committed:
Try larger drag handles on the spacer block

tofumatt reviewed (changes requested):
Tested locally and this works for me. I like it.

Once the SCSS is tidied up a bit I say we ship it.

karmatosed reviewed (approved):
I think this is a good addition, thanks!

chrisvanpatten commented on packages/block-library/src/spacer/editor.scss:
```diff
@@ -9,17 +8,32 @@
+	// Offset the handle's position
 	position: absolute;
-	background: theme(primary);
+	left: calc(50% - #{$resize-handler-container-size / 2}) !important;
```
@gwwar @tofumatt Agreed, but the [resize component](https://github.com/bokuweb/re-resizable) we're using here applies inline styles which we need to override. However I also see that they support a `style` prop so I'll see if I can just disable the inline styles.

chrisvanpatten commented on packages/block-library/src/spacer/editor.scss:
```diff
@@ -9,17 +8,32 @@
+	bottom: calc(#{$resize-handler-container-size / 2} * -1) !important;
+
+	// Apply the width/height of the container size plus padding so the handle
+	// sits in the middle of the grabbable area
```
I like nitpicks 😜  I'll get it updated! Agreed, a linting rule would be nice, but maybe this could be put into contributing docs or something too?

tofumatt replied:
I would approve such a pull request! 😄

chrisvanpatten committed:
Move resize styles to mixins, document use of !important, apply to image block

chrisvanpatten commented:
@tofumatt Unfortunately I was not able to remove the `!important`. I tried passing an empty object, `false`, and `null` to the `ResizableBox` component and the styles persisted.

So, in lieu of that, I wrote a bit of documentation and linked back to this ticket!

A few other tweaks:

+ Moved the container size variable into `edit-post/assets/stylesheets/_variables.scss`
+ Added mixins for the resize handle container and visible handle
+ Applied those mixins in the spacer block
+ Updated the image block to use the new mixins
+ Successfully tested the image block in various configurations (different alignments and such)

🚀

chrisvanpatten committed:
Use BEM-ish/more descriptive names for the mixins

tofumatt reviewed (commented):
😭 Arg, so it's an upstream issue.

Can we file/find/link to an upstream issue that would allow us to override/disable their styles so we can avoid the `!important` usage?

After that I would... probably begrudgingly accept this 😅

jasmussen commented:
Tried checking this out, and not really seeng any visual difference:

<img width="431" alt="screen shot 2018-10-04 at 14 09 56" src="https://user-images.githubusercontent.com/1204802/46493818-51465b80-c7df-11e8-9c25-2cd2b478bf0f.png">

<img width="294" alt="screen shot 2018-10-04 at 14 09 43" src="https://user-images.githubusercontent.com/1204802/46493828-54d9e280-c7df-11e8-826d-66bd4baea278.png">

Which is the point, of course. And in that vein, ship it!

The bottom drag handle does sort of sit at the same spot where the sibling inserter is, which could be better. But that's a separate issue.

chrisvanpatten committed:
Un-set inline styles on drag handles to avoid !important

chrisvanpatten committed:
Add some full stops, full stop.

tofumatt commented:
I know 235e8dc06f9c9f7576b533a29d0666c99b9ffd36 is more code but I super-appreciate the lack of `!important`s! ❤️❤️❤️

chrisvanpatten commented:
@tofumatt Okay I think I got it now! I couldn't remove the entire style definition, but I could remove the properties individually. So we can now avoid `!important` here!

I'm pretty sure spacers and images are the only uses of drag handles in Gutenberg core blocks but if I'm wrong definitely let me know 🤘

A future improvement here might be to ship our own `ResizableBox` that extends `re-resizable` with shared class names across all block implementations rather than each block having its own block-specific resize handle classname. We could also probably simplify the API a bit with a wrapper. But I don't think that's necessary to tackle here and would require some other decisions that I'm not qualified to make :)

chrisvanpatten commented:
@tofumatt You beat me by a hair ;-) But yes, I agree — no `!important` is worth it!

jasmussen commented:
> I'm pretty sure spacers and images are the only uses of drag handles in Gutenberg core blocks but if I'm wrong definitely let me know 🤘

They will also be used in https://github.com/WordPress/gutenberg/pull/9416 by @jorgefilipecosta

tofumatt reviewed (approved):
Sooooo, this is good but given we reproduce [this bit of code](https://github.com/WordPress/gutenberg/pull/10331/files#diff-f5fc3d5a78f58a5274bc196ebc657c5bR435) twice, maybe we want a common component that is a `<ResizableBox>` with the style overrides? I could see this happening again without someone catching it.

I think this is worth shipping now but can you file a follow-up issue to combine these? Actually, maybe it wouldn't make sense as the overrides vary? I dunno, if you think it's worth combining them a follow-up code quality issue would be cool. If you think it's not worth it that's fine.

chrisvanpatten commented:
@jasmussen Ooh, thanks for the heads up.

@tofumatt Yeah I think it's worth following this up with a "create our own wrapper around ResizableBox" PR, but I'd like to _handle_ (heh) that in a separate issue :)

chrisvanpatten commented:
Thanks for the support and guidance on this everyone!
